frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java

```diff
@@ -344,18 +338,19 @@ private ContainerRequest setupContainerRequest(TaskStatus taskStatus) throws Exc
           requestResource, maxResource);
     }
 
-    if (requestResource.getGpuNumber() > 0) {
-      updateNodeReports(yarnClient.getNodeReports(NodeState.RUNNING));
+    updateNodeReports(yarnClient.getNodeReports(NodeState.RUNNING));
+    SelectionResult selectionResult = selectionManager.select(requestResource, taskRoleName);
 
-      SelectionResult selectionResult = selectionManager.select(requestResource, requestNodeLabel, requestNodeGpuType);
-      if (selectionResult != null) {
-        ResourceDescriptor optimizedRequestResource = YamlUtils.deepCopy(requestResource, ResourceDescriptor.class);
-        optimizedRequestResource.setGpuAttribute(selectionResult.getGpuAttribute());
-        return HadoopUtils.toContainerRequest(optimizedRequestResource, requestPriority, null, selectionResult.getNodeHost());
-      }
+    ResourceDescriptor optimizedRequestResource = selectionResult.getOptimizedResource();
+    if (selectionResult.getSelectedNodeHosts().size() <= 0) {
+      return HadoopUtils.toContainerRequest(optimizedRequestResource, requestPriority, requestNodeLabel, null);
     }
 
-    return HadoopUtils.toContainerRequest(requestResource, requestPriority, requestNodeLabel, null);
+    //Random pick a host from the result set
```

Comment:
>Random [](start = 6, length = 6)

maybe we should explain why we need randomization here: to avoid conflicted request from concurrent container request (from different jobs)